### PR TITLE
Proprietary Serde type

### PR DESCRIPTION
**Pull request by svroonland**
Implements #30

* `Serde[T] extends Serializer[T] with Deserializer[T]`
* Internal `KafkaConsumer` and `KafkaProducer` works with `Array[Byte]` serdes for keys and values
* Instances for default Kafka serdes
* Creation from existing Kafka serdes
* Operators for effectful and non-effectful conversion of Serdes (mapping, contramapping and invariant mapping), can be used to construct new Serdes from existing ones
* `either` operator on `Deserializer` to explicitly handle deserialization failures.

### CODE REVIEW

**Comment by iravid** [comment]
Wow that was pretty quick @svroonland! :-)

Here are some thoughts:
- Parameterize the serdes with `R`, and allow access to the record metadata from the environment. So a `Deserializer[R, T]` returns a `RIO[R with RecordMetadata, T]`
- What do we gain from having the `Serde` trait rather than just separate Serializer/Deserializers?
- Deserializer can go up to flatMap, definitely useful

**Comment by iravid** [comment]
Oh and I'm really happy you got rid of `K` and `V` from the runloop 🙏🏻

**Comment by iravid** [comment]
Another thought: we definitely want a distinction between pure and effectful serializers/deserializers. This will let us shortcut the ZIO runloop if we're just working with pure deserializers.

**Comment by svroonland** [comment]
> Wow that was pretty quick @svroonland! :-)

Yeah, I have written these codec-like things before several times. It's easy when you treat them as invariant functors (in the category of endo-functors of course (that part is a joke)) :)

* I don't think I have yet mastered ZIO enough to see how that `RecordMetaData` thing would work, but would be happy to work on it with some pointers.
* I think it's nice to have a `Serde` type so you can define the serializer and deserializer for a new datatype as a single unit, by doing an `inmap` on an existing data type for example. 
* What do you mean exactly, go up to flatMap?

> Oh and I'm really happy you got rid of K and V from the runloop 

Indeed!

By the way, if you have a different design or implementation in mind, don't hesitate to disregard or cherry-pick from this PR! Or just tell me what to change :)

**Comment by svroonland** [comment]
Hmm, about effectful serdes: I don't think we should want arbitrary effects in serdes, maybe we should just use `Try[T]` as result type instead of `Task[T]`..

**Comment by iravid** [comment]
Effectful serdes are definitely useful! E.g., running logging effects. Don't give up on them :-)

**Comment by svroonland** [comment]
I see. But perhaps its simpler to model them non-effectful and if you want to do something like logging you do that in the stream consumption?

**Comment by svroonland** [comment]
.. To elaborate on that: you could get a stream of `ConsumerRecord[Array[Byte], Array[Byte]]` and do your own effectful deserialization there.

**Comment by iravid** [comment]
Yeah, but that's just introducing another effect type where we could just use ZIO and attain simpler code.

**Comment by iravid** [comment]
Maybe I should rephrase. Specifying serdes as `Array[Byte] => ZIO[R, Throwable, A]` has two main upsides:
1. We don't artificially constrain the user - after all, there's no reason not to allow them if we disregard our opinions of what should and shouldn't run in the serdes;
2. We attain simpler composition of the serdes with our stream, as we don't need to convert between an intermediate representation.

**Comment by svroonland** [comment]
Sounds fair. One use I can think of is some sort of Avro-schema lookup serializer, which needs to connect to a schema registry or something.

**Comment by svroonland** [comment]
Well, that seems to compile :)

I don't really like the `Consumer.make[Any, String, String]`. Since we can't infer the key and value types anyway, maybe it's better to change `Consumer.make` to

```scala
  def make[R, K, V](
    settings: ConsumerSettings,
    keyDeserializer: Deserializer[R, K],
    valueDeserializer: Deserializer[R, V]
)
```

so we can do `Consumer.make(settings, Serdes.string, Serdes.string)`.

**Comment by iravid** [comment]
@svroonland I actually suggest deferring the specification of a deserializer to the `partitionedStream` / `plainStream` methods.

**Comment by svroonland** [comment]
Updated PR. If we're happy with this implementation I can start to add some docs and some tests.

File: src/main/scala/zio/kafka/client/serde/Serde.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * A serializer and deserializer for values of type T
+ *
+ * @tparam R Environment available to the deserializer
+ * @tparam T Value type
+ */
+trait Serde[-R, T] extends Deserializer[R, T] with Serializer[R, T]
+
+object Serde {
```
Comment (iravid) [inline src/main/scala/zio/kafka/client/serde/Serde.scala]: Could you please add a constructor that wraps Kafka's `Serde`?

Comment (iravid) [reply src/main/scala/zio/kafka/client/serde/Serde.scala]: Oh, that's in Serdes. Why not move them here?

Comment (svroonland) [reply src/main/scala/zio/kafka/client/serde/Serde.scala]: That makes sense

**Pull request merged**